### PR TITLE
Add appointment to calendar 4171

### PR DESCRIPTION
**Pull request by vbahinwillit**
## Description
As a veteran, I want to be able to add healthcare appointments to my personal calendar so that I can easily keep track of them and be reminded of them.

## Testing done
Unit tested in Safari 13.0.2, Chrome 79.0.3945.79, and Outlook

## Screenshots
Screen Shot 2020-01-08 at 2 56 18 PM

## Acceptance criteria
- [ ]  'Add to calendar' link / button is displayed next to confirmed appointments
- [ ]  Clicking on link adds the event to a veteran's calendar

## Definition of done
- [ ] Events are logged appropriately
- [ ] Documentation has been updated, if applicable
- [ ] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [ ] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

### CODE REVIEW

File: src/applications/vaos/utils/appointment.jsx [hunk 1]
```diff
@@ -371,12 +372,12 @@ export function getAppointmentInstructions(appt) {
       return appt.instructionsToVeteran;
     case APPOINTMENT_TYPES.vaAppointment: {
       const bookingNotes =
-        appt.vdsAppointments?.[0]?.bookingNote ||
-        appt.vvsAppointments?.[0]?.bookingNotes;
+        appt.vdsAppointments[0]?.bookingNote ||
```
Comment (jbalboni) [inline src/applications/vaos/utils/appointment.jsx]: Why did you remove the `?.` after `vdsAppointments`? Those keeps us from throwing errors if an object without `vdsAppointments` is passed to this function.

File: src/applications/vaos/utils/appointment.jsx [hunk 2]
```diff
@@ -463,3 +464,126 @@ export function getAppointmentTypeHeader(appt) {
...
+}
+
+/**
+ * Function to get the appointment address.
+ *
+ * @param {*} appt
+ * @param {*} facility
+ * @returns 'undefined' if there is no address
+ */
+function getAppointmentAddress(appt, facility) {
```
Comment (jbalboni) [inline src/applications/vaos/utils/appointment.jsx]: I wonder if it makes more sense to say `getAppointmentLocation`? I think location is a better description for something can return "Video conference."

File: src/applications/vaos/utils/appointment.jsx [hunk 3]
```diff
@@ -463,3 +464,126 @@ export function getAppointmentTypeHeader(appt) {
...
+  const title = getAppointmentTypeHeader(appt);
+  const filename = `${title.replace(' ', '_')}.ics`;
+
+  function download(text) {
+    const element = document.createElement('a');
+    element.setAttribute(
+      'href',
+      `data:text/calendar;charset=utf8,${encodeURIComponent(text)}`,
+    );
+    element.setAttribute('download', filename);
```
Comment (jbalboni) [inline src/applications/vaos/utils/appointment.jsx]: I don't think think the `download` attribute works on IE11: https://caniuse.com/#feat=download. It looks like the code from static-pages doesn't handle this either, so might be something we need to look into in a separate ticket.

File: src/applications/vaos/utils/appointment.jsx [hunk 4]
```diff
@@ -463,3 +464,126 @@ export function getAppointmentTypeHeader(appt) {
...
+ *
+ * @export
+ * @param {*} appt
+ * @param {*} facility
+ */
+export function generateICS(appt, facility) {
+  const cal = new ICS.VCALENDAR();
+  const event = new ICS.VEVENT();
+
+  // TODO: Which one is better????
```
Comment (jbalboni) [inline src/applications/vaos/utils/appointment.jsx]: `appointmentTypeHeader` is better. The other one is actually unused now outside of tests and should be deleted.

File: src/applications/vaos/utils/appointment.jsx [hunk 5]
```diff
@@ -463,3 +464,126 @@ export function getAppointmentTypeHeader(appt) {
...
+}
+
+/**
+ * Function to generate ICS file suitable for import into Calendar.
+ *
+ * @export
+ * @param {*} appt
+ * @param {*} facility
+ */
+export function exportICSFile(appt, facility) {
```
Comment (jbalboni) [inline src/applications/vaos/utils/appointment.jsx]: I don't know how much you played around with the code in here, but it doesn't seem like this code around creating an element and clicking on it should be necessary. I wonder if we could just set the `href` on the link in `ConfirmedAppointmentListItem` to this `data:` string?

File: src/applications/vaos/components/ConfirmedAppointmentListItem.jsx [hunk 6]
```diff
@@ -97,8 +98,37 @@ export default function ConfirmedAppointmentListItem({
         )}
       </div>
 
+      {!allowCancel &&
+        !canceled && (
+          <div className="vads-u-margin-top--2">
+            <button
```
Comment (jbalboni) [inline src/applications/vaos/components/ConfirmedAppointmentListItem.jsx]: It might make sense to wrap this button in a separate `AddToCalendar` component. We'll also likely want to use it on the confirmation page.

File: src/applications/vaos/components/ConfirmedAppointmentListItem.jsx [hunk 7]
```diff
@@ -97,8 +98,37 @@ export default function ConfirmedAppointmentListItem({
...
+                {' '}
+                on {getAppointmentDate(appointment)}
+              </span>
+            </button>
+          </div>
+        )}
+
       {allowCancel && (
         <div className="vads-u-margin-top--2">
+          <button
```
Comment (jbalboni) [inline src/applications/vaos/components/ConfirmedAppointmentListItem.jsx]: I think we could avoid having two of these buttons in our component markup with some reworking of when the outer div is shown.

File: src/applications/vaos/components/ConfirmedAppointmentListItem.jsx [hunk 8]
```diff
@@ -97,8 +98,37 @@ export default function ConfirmedAppointmentListItem({
         )}
       </div>
 
+      {!allowCancel &&
+        !canceled && (
+          <div className="vads-u-margin-top--2">
+            <button
+              onClick={() => exportICSFile(appointment, facility)}
+              aria-label="App appointment to calendar"
```
Comment (jbalboni) [inline src/applications/vaos/components/ConfirmedAppointmentListItem.jsx]: I would guess you mean "Add" instead of "App"?

File: src/applications/vaos/components/ConfirmedAppointmentListItem.jsx [hunk 9]
```diff
@@ -97,8 +98,37 @@ export default function ConfirmedAppointmentListItem({
...
+              className="vaos-appts__cancel-btn va-button-link vads-u-margin--0 vads-u-flex--0"
+            >
+              Add to calendar
+              <span className="sr-only">
```
Comment (jbalboni) [inline src/applications/vaos/components/ConfirmedAppointmentListItem.jsx]: I don't think we need the sr only text here if we also have an `aria-label`. I would change `aria-label` to be `Add to calendar on {getAppointmentDate(appointment)}` and leave this button text as "Add to calendar".

File: src/applications/vaos/utils/appointment.jsx [hunk 10]
```diff
@@ -463,3 +464,126 @@ export function getAppointmentTypeHeader(appt) {
...
 }
+
+/**
+ * Function to get the appointment duration in minutes. The default is 60 minutes.
+ *
+ * @param {*} appt
+ * @returns
+ */
+function getAppointmentDuration(appt) {
+  const type = getAppointmentType(appt);
```
Comment (jbalboni) [inline src/applications/vaos/utils/appointment.jsx]: Would be good to write some unit tests for the new functions in this file.

File: src/applications/vaos/components/ConfirmedAppointmentListItem.jsx (hunk 8 above)
Comment (vbahinwillit) [reply src/applications/vaos/components/ConfirmedAppointmentListItem.jsx]: Done

File: src/applications/vaos/components/ConfirmedAppointmentListItem.jsx (hunk 9 above)
Comment (vbahinwillit) [reply src/applications/vaos/components/ConfirmedAppointmentListItem.jsx]: Done

File: src/applications/vaos/utils/appointment.jsx (hunk 4 above)
Comment (vbahinwillit) [reply src/applications/vaos/utils/appointment.jsx]: Done.

File: src/applications/vaos/utils/appointment.jsx (hunk 2 above)
Comment (vbahinwillit) [reply src/applications/vaos/utils/appointment.jsx]: The name 'getAppointmentLocation' is taken.

File: src/applications/vaos/utils/appointment.jsx (hunk 1 above)
Comment (vbahinwillit) [reply src/applications/vaos/utils/appointment.jsx]: Added the optional chaining operator back. However, I did find that this feature is not supported by all browsers.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining

Comment (jbalboni) [reply src/applications/vaos/utils/appointment.jsx]: All of our JS is run through Babel, which we use to transform JS with more modern features into JS that can be used by all the browsers we support. We have it set up to transform optional chaining (https://github.com/department-of-veterans-affairs/vets-website/blob/master/.babelrc#L32), so we should be good.

This is a good thing to think about whenever you're using new features, though. Things that are features of the "web platform" (like the `download` attribute I commented on below, or new browser apis) aren't always something we can handle through Babel.

File: src/applications/vaos/components/ConfirmedAppointmentListItem.jsx (hunk 7 above)
Comment (vbahinwillit) [reply src/applications/vaos/components/ConfirmedAppointmentListItem.jsx]: I'm not sure how since the "allowCancel" flag checks for appointments types not equal to CC and appointments not equal to video visit.

Comment (jbalboni) [reply src/applications/vaos/components/ConfirmedAppointmentListItem.jsx]: We'd probably have to tweaks some of those checks. I'm not sure of exactly what it would look like without playing around with it, so I don't think we should hold up this PR for it or anything.

File: src/applications/vaos/utils/appointment.jsx (hunk 3 above)
Comment (vbahinwillit) [reply src/applications/vaos/utils/appointment.jsx]: Agreeded

File: src/applications/vaos/components/ConfirmedAppointmentListItem.jsx (hunk 6 above)
Comment (vbahinwillit) [reply src/applications/vaos/components/ConfirmedAppointmentListItem.jsx]: Done

File: src/applications/vaos/components/ConfirmedAppointmentListItem.jsx (hunk 9 above)
Comment (vbahinwillit) [reply src/applications/vaos/components/ConfirmedAppointmentListItem.jsx]: Looks like the sr-only filters the anchor text displayed.  Without it, the link displays, "Add to calendar on November 7, 2020" which doesn't match the screenshots. If anything, the aria-label should be removed and allow the screen reader to just read the anchor text.

Comment (jbalboni) [reply src/applications/vaos/components/ConfirmedAppointmentListItem.jsx]: Ah, I meant that you could remove the date part from the link text. So the link should read "Add to calendar" and then aria-label would be "Add to calendar on November 7, 2020"

That keeps the link text short for sighted users, who can tell visually which date the link is associated with, but lets screen reader users get some extra context. We do a similar thing with the Edit links on the review page.

File: src/applications/vaos/utils/appointment.jsx (hunk 5 above)
Comment (vbahinwillit) [reply src/applications/vaos/utils/appointment.jsx]: Done.

File: src/applications/vaos/components/AddToCalendar.jsx [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+
+import {
+  getAppointmentTypeHeader,
+  getAppointmentDate,
+  generateICS,
+} from '../utils/appointment';
+
+export default class AddToCalendar extends React.Component {
+  render() {
+    const { appointment, facility } = this.props;
```
Comment (jbalboni) [inline src/applications/vaos/components/AddToCalendar.jsx]: I hesitate to say this given that you made 1k lines of unit test changes in another file, but we need some tests for this component. We can make a separate ticket for that.

File: src/applications/vaos/tests/utils/appointment.unit.spec.jsx [hunk 12]
```diff
@@ -67,134 +279,746 @@ describe('VAOS appointment helpers', () => {
...
+        facilityId: '983',
+        ccAppointmentRequest: {
+          preferredProviders: [{}],
+        },
+      });
+      expect(React.isValidElement(component)).to.be.true;
+    });
+  });
+
+  // NOTE: Can't do negative test since app assumes data is valid. This will result
```
Comment (jbalboni) [inline src/applications/vaos/tests/utils/appointment.unit.spec.jsx]: @vbahinwillit Can you create an issue for this?

**Review by jbalboni — APPROVED** [review]
LGTM. I made a comment about some unit tests for the new component, we can do those in a separate ticket. The additional unit tests are definitely an improvement, but in the future we need to make sure that we break out changes like that into separate PRs and try to keep them down to a reasonable size (300-500 lines).

File: src/applications/vaos/components/AddToCalendar.jsx (hunk 11 above)
Comment (vbahinwillit) [reply src/applications/vaos/components/AddToCalendar.jsx]: Ok.

File: src/applications/vaos/components/ConfirmedAppointmentListItem.jsx (hunk 7 above)
Comment (vbahinwillit) [reply src/applications/vaos/components/ConfirmedAppointmentListItem.jsx]: Ok.

File: src/applications/vaos/components/ConfirmedAppointmentListItem.jsx (hunk 9 above)
Comment (vbahinwillit) [reply src/applications/vaos/components/ConfirmedAppointmentListItem.jsx]: Done

File: src/applications/vaos/components/AddToCalendar.jsx (hunk 11 above)
Comment (vbahinwillit) [reply src/applications/vaos/components/AddToCalendar.jsx]: Ticket created. #4890

**Review by bkjohnson — APPROVED** [review]
Thank you for the tests.
